src/Studio/Widgets/ChannelMultiSelectionWidget.h

```diff
@@ -30,77 +30,411 @@
+    void connectedChanged(bool connected);
+
+    void currentServiceChanged(int currentService);
+
+private slots:
+    //QBluetothDeviceDiscoveryAgent
+    void addDevice(const QBluetoothDeviceInfo&);
+    void onScanFinished();
+    void onDeviceScanError(QBluetoothDeviceDiscoveryAgent::Error);
+
+    //QLowEnergyController
+    void onServiceDiscovered(const QBluetoothUuid&);
+    void onServiceScanDone();
+    void onControllerError(QLowEnergyController::Error);
+    void onDeviceConnected();
+    void onDeviceDisconnected();
+
+    //QLowEnergyService
+    void onServiceStateChanged(QLowEnergyService::ServiceState s);
+    void onCharacteristicChanged(const QLowEnergyCharacteristic& c,
+        const QByteArray& value);
+    void serviceError(QLowEnergyService::ServiceError e);
+
+    void read();
+    void onCharacteristicRead(const QLowEnergyCharacteristic& c, const QByteArray& value);
+    void onCharacteristicWrite(const QLowEnergyCharacteristic& c, const QByteArray& value);
+    void update_currentService(int currentSerice);
+
+private:
+    inline void waitForWrite();
+    void update_connected(bool connected) {
+        if (connected != m_connected) {
+            m_connected = connected;
+            emit connectedChanged(connected);
+        }
+    }
+
+    QBluetoothDeviceDiscoveryAgent* m_deviceDiscoveryAgent;
+    QLowEnergyDescriptor m_notificationDesc;
+    QLowEnergyController* m_control;
+    QList<QBluetoothUuid> m_servicesUuid;
+    QLowEnergyService* m_service;
+    QLowEnergyCharacteristic m_readCharacteristic;
+    QLowEnergyCharacteristic m_writeCharacteristic;
+    QList<DeviceInfo*> m_devices;
+    //    bool m_foundService;
+    QTimer* m_readTimer;
+    bool m_connected;
+    void searchCharacteristic();
+    int m_currentService;
+    QLowEnergyService::WriteMode m_writeMode;
+};
 
 class ChannelMultiSelectionWidget : public QWidget
 {
-	Q_OBJECT
-	public:
-		// device-change behaviour
-		enum AutoSelectType { 
-			SELECT_NONE, 
-			SELECT_FIRST, 
-			SELECT_ALL 
-		};
+    Q_OBJECT
+
+        /*Q_PROPERTY(bool connected READ isConnected NOTIFY connectedChanged)
+        Q_PROPERTY(int currentService READ currentService WRITE setCurrentService NOTIFY currentServiceChanged)
+        QML_WRITABLE_PROPERTY(int, currentDevice)
+        QML_READONLY_PROPERTY(QStringList, devicesNames)
+        QML_READONLY_PROPERTY(QStringList, services)*/
+
+public:
+    // device-change behaviour
+    enum AutoSelectType {
+        SELECT_NONE,
+        SELECT_FIRST,
+        SELECT_ALL
+    };
+
+    // constructor & destructor
+    ChannelMultiSelectionWidget(QWidget* parent = NULL);
+    virtual ~ChannelMultiSelectionWidget();
+
+    // initialize
+    void Init();
+    void ReInit(Device* device = NULL);
+    void Scan_BLE();
+    void On_connect();
+    void On_Start();
+
+    // TODO: replace with transparent accessors (GetNumAvailableChannels(), GetAvailableChannel(uint32 index) etc.)
```

Comment:
I like these changes potentially @cyberjunk

Reply:
@dioptre, @cyberjunk, please complete the review process so that I can update the further.